plugins/tpm/fu-tpm-device.c

```diff
@@ -0,0 +1,291 @@
+	fu_device_add_child (FU_DEVICE (self), dev);
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_tpm_get_uint32 (ESYS_CONTEXT *ctx, guint32 query, guint32 *val, GError **error)
+{
+	TSS2_RC rc;
+	g_autofree TPMS_CAPABILITY_DATA *capability = NULL;
+
+	g_return_val_if_fail (val != NULL, FALSE);
+
+	rc = Esys_GetCapability (ctx, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
+				 TPM2_CAP_TPM_PROPERTIES, query, 1, NULL, &capability);
+	if (rc != TSS2_RC_SUCCESS) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+			     "capability request failed for query %x", query);
+		return FALSE;
+	}
+	if (capability->data.tpmProperties.count == 0) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+			     "no properties returned for query %x", query);
+		return FALSE;
+	}
+	if (capability->data.tpmProperties.tpmProperty[0].property != query) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+			     "wrong query returned (got %x expected %x)",
+			     capability->data.tpmProperties.tpmProperty[0].property,
+			     query);
+		return FALSE;
+	}
+
+	*val = GUINT32_FROM_BE (capability->data.tpmProperties.tpmProperty[0].value);
+	return TRUE;
+}
+
+static gchar *
+fu_plugin_tpm_get_string (ESYS_CONTEXT *ctx, guint32 query, GError **error)
+{
+	guint32 val = 0;
+	gchar result[5] = {'\0'};
+
+	/* return four bytes */
+	if (!fu_plugin_tpm_get_uint32 (ctx, query, &val, error))
+		return NULL;
+	memcpy (result, (gchar *) &val, 4);
+
+	/* convert non-ASCII into spaces */
+	for (guint i = 0; i < 4; i++) {
+		if (!g_ascii_isgraph (result[i]) && result[i] != '\0')
+			result[i] = 0x20;
+	}
+
+	return fu_common_strstrip (result);
+}
+
+static const gchar *
+fu_tpm_device_convert_manufacturer (const gchar *manufacturer)
+{
+	if (g_strcmp0 (manufacturer, "AMD") == 0)
+		return "AMD";
+	if (g_strcmp0 (manufacturer, "ATML") == 0)
+		return "Atmel";
+	if (g_strcmp0 (manufacturer, "BRCM") == 0)
+		return "Broadcom";
+	if (g_strcmp0 (manufacturer, "IBM") == 0)
+		return "IBM";
+	if (g_strcmp0 (manufacturer, "IFX") == 0)
+		return "Infineon";
+	if (g_strcmp0 (manufacturer, "INTC") == 0)
+		return "Intel";
+	if (g_strcmp0 (manufacturer, "LEN") == 0)
+		return "Lenovo";
+	if (g_strcmp0 (manufacturer, "NSM") == 0)
+		return "National Semiconductor";
+	if (g_strcmp0 (manufacturer, "NTZ") == 0)
+		return "Nationz";
+	if (g_strcmp0 (manufacturer, "NTC") == 0)
+		return "Nuvoton Technology";
+	if (g_strcmp0 (manufacturer, "QCOM") == 0)
+		return "Qualcomm";
+	if (g_strcmp0 (manufacturer, "SMSC") == 0)
+		return "SMSC";
+	if (g_strcmp0 (manufacturer, "STM") == 0)
+		return "ST Microelectronics";
+	if (g_strcmp0 (manufacturer, "SMSN") == 0)
+		return "Samsung";
+	if (g_strcmp0 (manufacturer, "SNS") == 0)
+		return "Sinosun";
+	if (g_strcmp0 (manufacturer, "TXN") == 0)
+		return "Texas Instruments";
+	if (g_strcmp0 (manufacturer, "WEC") == 0)
+		return "Winbond";
+	if (g_strcmp0 (manufacturer, "ROCC") == 0)
+		return "Fuzhou Rockchip";
+	return NULL;
```

Comment:
Aren't you missing Google?  `GOOG`

Reply:
Ahh, I was using the old `Vendor_ID_Registry_0-8_clean.pdf` -- I've added HPE, Microsoft and Google to the list. Good catch!